deepmd/utils/tabulate.py

```diff
@@ -0,0 +1,457 @@
+            ll = np.min(lower)
+            xx = np.arange(ll, uu, stride0, dtype=self.data_type)
+            xx = np.append(
+                xx,
+                np.arange(uu, extrapolate * uu, stride1, dtype=self.data_type),
+            )
+            xx = np.append(xx, np.array([extrapolate * uu], dtype=self.data_type))
+            nspline = ((uu - ll) / stride0 + (extrapolate * uu - uu) / stride1).astype(
+                int
+            )
+            self._build_lower(
+                "filter_net", xx, 0, uu, ll, stride0, stride1, extrapolate, nspline
+            )
+        elif self.descrpt_type == "A":
+            for ii in range(self.table_size):
+                if (self.type_one_side and not self._all_excluded(ii)) or (
+                    not self.type_one_side
+                    and (ii // self.ntypes, ii % self.ntypes) not in self.exclude_types
+                ):
+                    if self.type_one_side:
+                        net = "filter_-1_net_" + str(ii)
+                        # upper and lower should consider all types which are not excluded and sel>0
+                        idx = [
+                            (type_i, ii) not in self.exclude_types
+                            and self.sel_a[type_i] > 0
+                            for type_i in range(self.ntypes)
+                        ]
+                        uu = np.max(upper[idx])
+                        ll = np.min(lower[idx])
+                    else:
+                        ielement = ii // self.ntypes
+                        net = (
+                            "filter_" + str(ielement) + "_net_" + str(ii % self.ntypes)
+                        )
+                        if self.is_pt:
+                            uu = np.max(upper[ielement])
+                            ll = np.min(lower[ielement])
+                        else:
+                            uu = upper[ielement]
+                            ll = lower[ielement]
+                    xx = np.arange(ll, uu, stride0, dtype=self.data_type)
+                    xx = np.append(
+                        xx,
+                        np.arange(uu, extrapolate * uu, stride1, dtype=self.data_type),
+                    )
+                    xx = np.append(
+                        xx, np.array([extrapolate * uu], dtype=self.data_type)
+                    )
+                    nspline = (
+                        (uu - ll) / stride0 + (extrapolate * uu - uu) / stride1
+                    ).astype(int)
+                    self._build_lower(
+                        net, xx, ii, uu, ll, stride0, stride1, extrapolate, nspline
+                    )
+        elif self.descrpt_type == "T":
+            xx_all = []
+            for ii in range(self.ntypes):
+                if self.is_pt:
+                    uu = np.max(upper[ii])
+                    ll = np.min(lower[ii])
+                else:
+                    ll = lower[ii]
+                    uu = upper[ii]
```

Comment:
Could you explain why TF and PT are different here? It is also useful to add a comment into the code.

Reply:
When in descrpt T, although the execution paths of tf and pt in tabulate are exactly the same, the shapes of upper and lower obtained are not consistent. The pt version is a two-dimensional array. During my test, all the values ​​in each row of pt version are the same. But I still used the `min` and `max` to get porper upper and lower.